File: app/src/main/java/com/security/passwordmanager/ui/home/PasswordListFragment.java
package com.security.passwordmanager.ui.home;

import android.annotation.SuppressLint;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.security.passwordmanager.Cryptographer;
import com.security.passwordmanager.Data;
import com.security.passwordmanager.DataLab;
import com.security.passwordmanager.PasswordInfoActivity;
import com.security.passwordmanager.R;
import com.security.passwordmanager.Support;

import java.util.List;

public class PasswordListFragment extends Fragment {

    private RecyclerView mRecyclerView;
    private PasswordAdapter mAdapter;

    private Support mSupport;
    private DataLab mDataLab;
    private List<Data> dataList;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        View root = inflater.inflate(R.layout.fragment_password_list, container, false);
        mRecyclerView = root.findViewById(R.id.main_recycler_view);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        updateUI();

        return root;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        mSupport = Support.get(getActivity());
        mDataLab = DataLab.get(getActivity());
    }

    @Override
    public void onResume() {
        super.onResume();
        updateUI();
    }

    private void updateUI() {
        dataList = mDataLab.getDataList();

        if (mAdapter == null) {
            mAdapter = new PasswordAdapter();
            mRecyclerView.setAdapter(mAdapter);
        }
        else mAdapter.notifyDataSetChanged();
    }



    private class PasswordAdapter extends RecyclerView.Adapter<PasswordHolder> {

        private static final int OPENED_VIEW = 0xAAA;
        private int openedView = -1;

        @Override
        public int getItemViewType(int position) {
            if (position == openedView)
                return OPENED_VIEW;
            return super.getItemViewType(position);
        }

        @NonNull
        @Override
        public PasswordHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(getActivity());
            View view = inflater
                    .inflate(R.layout.list_item_main_text_view, parent, false);

            RecyclerView infoRecyclerView = view.findViewById(R.id.list_item_recycler_view_more_info);

            if (viewType == OPENED_VIEW)
                infoRecyclerView.setVisibility(View.VISIBLE);
            else
                infoRecyclerView.setVisibility(View.GONE);

            return new PasswordHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull PasswordHolder holder, int position) {
            Data data = dataList.get(position);
            holder.bindPassword(mDataLab.getAccountList(data.getAddress()));

            holder.setOnClickListener(v -> {
                if (openedView == position)
                    openedView = -1;
                else {
                    int oldOpen = openedView;
                    openedView = position;
                    notifyItemChanged(oldOpen);
                }
                notifyItemChanged(position);
            });
        }

        @Override
        public int getItemCount() {
            return dataList.size();
        }
    }



    private class PasswordHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final RecyclerView recyclerView;
        private final LinearLayout text_view;
        private final TextView textView_name, textView_url;
        private final Button button_more;

        private List<Data> mAccountList;

        public PasswordHolder(@NonNull View itemView) {
            super(itemView);
            text_view = itemView.findViewById(R.id.list_item_text_view);
            textView_name = itemView.findViewById(R.id.list_item_text_view_name);
            textView_url = itemView.findViewById(R.id.list_item_text_view_url);
            button_more = itemView.findViewById(R.id.list_item_button_more);

            recyclerView = itemView.findViewById(R.id.list_item_recycler_view_more_info);
            recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));


            button_more.setOnClickListener(this);
        }

        public void bindPassword(List<Data> accountList) {
            this.mAccountList = accountList;

            recyclerView.setAdapter(new MoreInfoAdapter(mAccountList));

            textView_name.setText(accountList.get(0).getNameWebsite());
            textView_url.setText(accountList.get(0).getAddress());

            textView_name.setBackgroundColor(mSupport.getBackgroundColor());
            textView_url.setBackgroundColor(mSupport.getBackgroundColor());
            button_more.setBackgroundTintList(ColorStateList.valueOf(mSupport.getFontColor()));

            textView_name.setTextColor(mSupport.getFontColor());
        }

        @SuppressLint("NonConstantResourceId")
        @Override
        public void onClick(View v) {
            startActivity(PasswordInfoActivity.newIntent(getContext(), mAccountList.get(0), PasswordInfoActivity.TYPE_DATA));
        }

        public void setOnClickListener(View.OnClickListener listener) {
            text_view.setOnClickListener(listener);
        }
    }



    private class MoreInfoAdapter extends RecyclerView.Adapter<MoreInfoHolder> {

        private final List<Data> accountList;

        public MoreInfoAdapter(List<Data> accountList) {
            this.accountList = accountList;
        }

        @NonNull
        @Override
        public MoreInfoHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(getActivity())
                    .inflate(R.layout.list_item_more_info, parent, false);
            return new MoreInfoHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull MoreInfoHolder holder, int position) {
            Data data = accountList.get(position);
            holder.bindInfo(data);
        }

        @Override
        public int getItemCount() {
            return accountList.size();
        }
    }



    private class MoreInfoHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final LinearLayout moreInfo;
        private final TextView accountName, login, password, comment;
        private final TextView head_login, head_password, head_comment;
        private final Button button_open_url;
        private final ImageButton copy;
        private Data data;

        public MoreInfoHolder(@NonNull View itemView) {
            super(itemView);
            moreInfo = itemView.findViewById(R.id.list_item_linear_layout_more);

            accountName = itemView.findViewById(R.id.list_item_text_view_name_of_account);

            login = itemView.findViewById(R.id.list_item_text_view_login);
            password = itemView.findViewById(R.id.list_item_text_view_password);
            comment = itemView.findViewById(R.id.list_item_text_view_comment);
            button_open_url = itemView.findViewById(R.id.list_item_button_open_url);

            head_login = itemView.findViewById(R.id.list_item_text_view_login_head);
            head_password = itemView.findViewById(R.id.list_item_text_view_password_head);
            head_comment = itemView.findViewById(R.id.list_item_text_view_comment_head);

            copy = itemView.findViewById(R.id.list_item_button_copy_login);
            copy.setOnClickListener(this);

            button_open_url.setTextColor(Color.WHITE);
        }

        public void bindInfo(Data data) {
            this.data = data;

            if (data.getComment().length() == 0) {
                comment.setVisibility(View.GONE);
                head_comment.setVisibility(View.GONE);
            }

            accountName.setText(data.getNameAccount());
            login.setText(data.getLogin());
            password.setText(data.getPassword());
            comment.setText(data.getComment());

            moreInfo.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            button_open_url.setBackgroundResource(mSupport.getButtonRes());

            accountName.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            login.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            password.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            comment.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            copy.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            copy.setImageTintList(ColorStateList.valueOf(mSupport.getFontColor()));

            accountName.setTextColor(mSupport.getDarkerGrayColor());
            login.setTextColor(mSupport.getFontColor());
            password.setTextColor(mSupport.getFontColor());
            comment.setTextColor(mSupport.getFontColor());

            head_login.setTextColor(mSupport.getFontColor());
            head_password.setTextColor(mSupport.getFontColor());
            head_comment.setTextColor(mSupport.getFontColor());
        }

        @SuppressLint("NonConstantResourceId")
        @Override
        public void onClick(View v) {
            switch (v.getId()) {
                case R.id.list_item_button_open_url:
                    String address;
                    if (data.getAddress().contains("www."))
                        address = "https://" + data.getAddress();
                    else if (data.getAddress().contains("https://www.") || data.getAddress().contains("http://www."))
                        address = data.getAddress();
                    else
                        address = "https://www." + data.getAddress();

                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(address));
                    startActivity(intent);
                    break;

                case R.id.list_item_button_copy_login:
                    ClipboardManager clipboard = (ClipboardManager)
                            requireActivity().getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText("", data.getLogin());
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(getActivity(), "Скопированно!", Toast.LENGTH_SHORT).show();
                    break;
            }
        }
    }
}

File: app/src/main/java/com/security/passwordmanager/ui/main/PasswordListFragment.kt
package com.security.passwordmanager.ui.main

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.text.InputType
import android.view.*
import android.webkit.URLUtil
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.security.passwordmanager.*
import com.security.passwordmanager.data.*
import com.security.passwordmanager.databinding.*
import com.security.passwordmanager.settings.SettingsViewModel
import com.security.passwordmanager.ui.account.PasswordActivity
import com.security.passwordmanager.ui.bank.BankCardActivity

class PasswordListFragment : Fragment() {

    companion object {
        private const val OPENED_VIEW_KEY = "opened_view"
//        private const val VISIBILITIES_KEY = "visibilities"
    }

    private lateinit var binding: FragmentPasswordListBinding

    private var adapter: PasswordAdapter? = null
    private var recyclerCallback: RecyclerCallback? = null
    private lateinit var searchView: SearchView

    private var openedView = -1

    private lateinit var settings: SettingsViewModel
    private lateinit var dataViewModel: DataViewModel
    private lateinit var dataList: List<Data>

    private fun openBottomSheet(bottomDialogFragment: ActionBottomDialogFragment) =
        activity?.supportFragmentManager?.let { bottomDialogFragment.show(it) }


    override fun onAttach(context: Context) {
        super.onAttach(context)
        recyclerCallback = activity as RecyclerCallback
    }

    override fun onDetach() {
        super.onDetach()
        recyclerCallback = null
    }


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        binding = FragmentPasswordListBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.main, menu)

        searchView = menu.findItem(R.id.menu_item_search).actionView as SearchView
        drawSearchView()

        searchView.setOnQueryTextListener(
            object: SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String) = false

                @SuppressLint("NotifyDataSetChanged")
                override fun onQueryTextChange(newText: String): Boolean {
                    dataList = dataViewModel.searchData(newText)
                    openedView =
                        if (dataList.size == 1) 0
                        else -1
                    adapter?.notifyDataSetChanged()
                    return true
                }
            })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setHasOptionsMenu(true)

        activity?.let {
            settings = SettingsViewModel.getInstance(this)
            dataViewModel = ViewModelProvider(it)[DataViewModel::class.java]
        }

        openedView = savedInstanceState.getInt(OPENED_VIEW_KEY, -1)

        binding.mainRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                recyclerCallback?.onScroll(
                    when {
                        dy > 1 -> RecyclerDirection.DOWN
                        dy < 1 -> RecyclerDirection.UP
                        else -> RecyclerDirection.STOPPED
                    },
                    RecyclerState.getState(recyclerView.scrollState)
                )

                super.onScrolled(recyclerView, dx, dy)
            }

            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                recyclerCallback?.onStateChanged(RecyclerState.getState(newState))

                super.onScrollStateChanged(recyclerView, newState)
            }
        })
    }


    @SuppressLint("NotifyDataSetChanged")
    override fun onResume() {
        super.onResume()
        dataList = dataViewModel.getDataList()

        if (adapter == null) {
            adapter = PasswordAdapter()
            binding.mainRecyclerView.adapter = adapter
        }
        else
            adapter?.notifyDataSetChanged()
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(OPENED_VIEW_KEY, openedView)
    }

    private fun drawSearchView() {
        val searchText = searchView.findViewById<EditText>(androidx.appcompat.R.id.search_src_text)
        val searchImage = searchView.findViewById<ImageView>(androidx.appcompat.R.id.search_button)
        val clearView = searchView.findViewById<ImageView>(R.id.search_close_btn)

        searchText.setTextColor(Color.WHITE)
        searchText.backgroundTintList = ColorStateList.valueOf(Color.WHITE)
        searchImage.imageTintList = ColorStateList.valueOf(Color.WHITE)
        clearView.imageTintList = ColorStateList.valueOf(Color.WHITE)
    }



    private inner class PasswordAdapter : RecyclerView.Adapter<PasswordHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PasswordHolder {
            val itemBinding = MainTextViewBinding
                .inflate(LayoutInflater.from(parent.context), parent, false)

            val holder = PasswordHolder(itemBinding)

            holder.setOnClickListener {
                if (holder.adapterPosition == openedView)
                    openedView = -1
                else {
                    val oldOpen = openedView
                    openedView = holder.adapterPosition
                    notifyItemChanged(oldOpen)
                }
                notifyItemChanged(holder.adapterPosition)
            }

            return holder
        }

        override fun onBindViewHolder(holder: PasswordHolder, position: Int) {
            val data = dataList[position]
            holder.bindPassword(data, position == openedView)
        }

        override fun getItemCount() = dataList.size
    }




    private inner class PasswordHolder(private val itemBinding: MainTextViewBinding):
        RecyclerView.ViewHolder(itemBinding.root), View.OnClickListener {

        private lateinit var accountList: ArrayList<Data>
        private var moreInfoAdapter: MoreInfoAdapter? = null

        private lateinit var data: Data

        init {
            itemBinding.buttonMore.setOnClickListener(this)
        }


        private fun LayoutMainHeaderBinding.getTextFields() = Pair(heading, subtitle)

        private fun LayoutMainHeaderBinding.setText(headingText: String, subtitleText: String) {
            val fields = getTextFields()
            fields[0].text = headingText
            fields[1].text = subtitleText
            fields.setAll { it.setBackgroundColor(settings.backgroundColor) }
            fields[0].setTextColor(settings.fontColor)
        }

        private fun LayoutMainHeaderBinding.getText() = Pair {
            getTextFields()[it].text.toString()
        }


        fun bindPassword(data: Data, isShown: Boolean) {
            this.data = data
            accountList = dataViewModel.getAccountList(data) as ArrayList<Data>

            moreInfoAdapter = MoreInfoAdapter(accountList)
            itemBinding.recyclerViewMoreInfo.adapter = moreInfoAdapter

            updateArrow(isShown)
            setVisible(isShown)

            when (data) {
                is Website -> itemBinding.textView.setText(data.nameWebsite, data.address)
                is BankCard -> itemBinding.textView.setText(data.bankName, data.cardNumber)
            }

            itemBinding.imageView.imageTintList = ColorStateList.valueOf(settings.headerColor)
            itemBinding.buttonMore.imageTintList = ColorStateList.valueOf(settings.fontColor)
            itemBinding.buttonMore.setBackgroundColor(settings.backgroundColor)
        }

        private fun setVisible(visible: Boolean) {
            val visibility = if (visible) View.VISIBLE else View.GONE
            itemBinding.recyclerViewMoreInfo.visibility = visibility
        }

        override fun onClick(v: View?) {
            /** v = btnMore */
            val text = itemBinding.textView.getText()

            val bottomDialogFragment = ActionBottom.newInstance(activity as AppCompatActivity)
            bottomDialogFragment.setHeading(text[0], text[1])

            bottomDialogFragment.addView(R.drawable.edit, R.string.edit) {
                when(data) {
                    is Website -> startActivity(
                        PasswordActivity.getIntent(context, data.key))
                    is BankCard -> startActivity(
                        BankCardActivity.getIntent(context, data.key))
                }
                bottomDialogFragment.dismiss()
            }

            bottomDialogFragment.addView(R.drawable.copy, R.string.copy_info) {
                dataViewModel.copyAccountList(accountList)
                bottomDialogFragment.dismiss()
            }

            bottomDialogFragment.addView(R.drawable.delete, R.string.delete_password) {
                dataViewModel.deleteRecords(data)

                bottomDialogFragment.dismiss()
                adapter?.notifyItemRemoved(adapterPosition)
            }
            openBottomSheet(bottomDialogFragment)
        }

        fun setOnClickListener(l: View.OnClickListener) =
            itemBinding.mainLayout.setOnClickListener(l)


        //true - стрелка вниз, false - вправо
        private fun updateArrow(isShown: Boolean) = if (isShown) {
            itemBinding.motionContainer.transitionToEnd()
        }
        else {
            itemBinding.motionContainer.transitionToStart()
        }




        private inner class MoreInfoAdapter(private val accountList: ArrayList<Data>):
            RecyclerView.Adapter<MoreInfoHolder>() {

            override fun getItemViewType(position: Int) =
                accountList[position].type.number

            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MoreInfoHolder {
                val layoutInflater = LayoutInflater.from(context)

                val moreInfoBinding = when(viewType) {
                    DataType.BANK_CARD.number -> MoreBankCardBinding
                        .inflate(layoutInflater, parent, false)

                    else -> MoreWebsiteBinding
                        .inflate(layoutInflater, parent, false)
                }

                return MoreInfoHolder(moreInfoBinding)
            }

            override fun onBindViewHolder(holder: MoreInfoHolder, position: Int) {
                val data = accountList[position]
                holder.bindInfo(data, position)
            }

            override fun getItemCount() =
                accountList.size
        }



        private inner class MoreInfoHolder(private val moreInfoBinding: ViewBinding):
            RecyclerView.ViewHolder(moreInfoBinding.root) {

            private lateinit var data: Data
            private var pos = 0

            private val bottomDialogFragment = ActionBottom.newInstance(activity as AppCompatActivity)

            private var isPasswordVisible = false

            fun TextView.setTextColor() =
                setTextColor(settings.fontColor)

            fun ImageButton.setOnCopyListener(text: String) {
                setOnClickListener { dataViewModel.copyText(text) }
                setBackgroundColor(settings.layoutBackgroundColor)
                imageTintList = ColorStateList.valueOf(settings.fontColor)
            }

            fun TextViewFieldBinding.setText(text: String) {
                textView.text = text
            }


            init {
                when (moreInfoBinding) {
                    is MoreWebsiteBinding -> {
                        moreInfoBinding.buttonOpenUrl.setOnClickListener {
                            openUrl()
                        }

                        moreInfoBinding.password.buttonVisibility.show()
                        moreInfoBinding.password.buttonVisibility.setOnClickListener {
                            updatePasswordView()
                        }

                        updatePasswordView(isPasswordVisible)
                    }
                    is MoreBankCardBinding -> {
                        //todo доделать bankCard binding
                    }
                }
            }

            fun bindInfo(data: Data, position: Int) {
                this.data = data
                this.pos = position

                when (moreInfoBinding) {
                    is MoreWebsiteBinding -> {
                        if (data !is Website) return

                        moreInfoBinding.run {
                            login.setText(data.login)
                            password.setText(data.password)
                            comment.setText(data.comment)

                            if (data.nameAccount.isEmpty())
                                nameAccount.hide()
                            else
                                nameAccount.txt = data.nameAccount

                            if (data.comment.isEmpty()) {
                                comment.textView.hide()
                                comment.fieldItemButtonCopy.hide()
                                commentHead.hide()
                            }

                            createBottomSheet()

                            //colors
                            root.backgroundTintList =
                                ColorStateList.valueOf(settings.layoutBackgroundColor)

                            for (item in arrayOf(login, password, comment)) {
                                item.textView.setBackgroundColor(settings.layoutBackgroundColor)
                                item.textView.setTextColor()
                            }

                            login.fieldItemButtonCopy.setOnCopyListener(data.login)
                            password.fieldItemButtonCopy.setOnCopyListener(data.password)
                            comment.fieldItemButtonCopy.setOnCopyListener(data.comment)

                            loginHead.setTextColor()
                            passwordHead.setTextColor()
                            commentHead.setTextColor()

                            nameAccount.setBackgroundColor(settings.layoutBackgroundColor)

                            password.buttonVisibility.setBackgroundColor(settings.layoutBackgroundColor)

                            password.buttonVisibility.imageTintList =
                                ColorStateList.valueOf(settings.fontColor)

                            buttonOpenUrl.setBackgroundResource(settings.buttonRes)

                            root.setOnClickListener {
                                if (settings.baseSettings.isUsingBottomView)
                                    bottomDialogFragment.show(childFragmentManager)
                            }
                        }
                    }
                }
            }


            fun updatePasswordView() {
                isPasswordVisible = !isPasswordVisible
                updatePasswordView(isPasswordVisible)
            }

            fun updatePasswordView(visibility: Boolean) =
                (moreInfoBinding as MoreWebsiteBinding).apply {
                    if (visibility) {
                        password.textView.inputType = InputType.TYPE_CLASS_TEXT
                        password.buttonVisibility.setImageResource(R.drawable.visibility_off)
                        password.buttonVisibility.contentDescription =
                            getString(R.string.hide_password)
                    } else {
                        password.textView.inputType = 129
                        password.buttonVisibility.setImageResource(R.drawable.visibility_on)
                        password.buttonVisibility.contentDescription =
                            getString(R.string.show_password)
                    }
                }

            //todo нужно ли в bankCard?
            fun createBottomSheet() {
                moreInfoBinding as MoreWebsiteBinding

                val heading = if (moreInfoBinding.nameAccount.isEmpty()) {
                    (data as Website).login
                }
                else moreInfoBinding.nameAccount.text.toString()

                bottomDialogFragment.setHeading(heading, beautifulDesign = true)

                bottomDialogFragment.addView(R.drawable.edit, R.string.edit) {
                    bottomDialogFragment.dismiss()
                    startActivity(PasswordActivity
                        .getIntent(context, (data as Website).address, pos))
                }
            }

            fun openUrl() {
                val website = data as Website

                val address = when {
                    "www." in website.address -> "https://${website.address}"
                    "https://www." in website.address ||
                            "http://www." in website.address -> website.address
                    else -> "https://www.${website.address}"
                }

                if (URLUtil.isValidUrl(address)) {
                    val intent = Intent(Intent.ACTION_VIEW, Uri.parse(address))
                    startActivity(intent)
                } else
                    showToast(activity, "Неправильный адрес!")
            }
        }
    }
}